allspice/utils/list_components.py

```diff
@@ -0,0 +1,114 @@
+import time
+
+from allspice.exceptions import NotYetGeneratedException
+from ..allspice import AllSpice
+from ..apiobject import Repository
+
+PCB_FOOTPRINT_ATTR_NAME = "PCB Footprint"
+PART_REFERENCE_ATTR_NAME = "Part Reference"
+
+
+def list_components_for_orcad(
+    allspice_client: AllSpice,
+    repository: Repository,
+    dsn_path: str,
+    ref: str = "main",
+) -> list[dict[str, str]]:
+    """
+    Get a list of all components in an OrCAD DSN schematic.
+
+    :param client: An AllSpice client instance.
+    :param repository: The repository containing the OrCAD schematic.
+    :param dsn_path: The path to the OrCAD DSN file from the repo root. For
+        example, if the schematic is in the folder "Schematics" and the file
+        is named "example.dsn", the path would be "Schematics/example.dsn".
+    :param ref: Optional git ref to check. This can be a commit hash, branch
+        name, or tag name. Default is "main", i.e. the main branch.
+    :return: A list of all components in the OrCAD schematic. Each component is
+        a dictionary with the keys being the attributes of the component and the
+        values being the values of the attributes. A `_name` attribute is added
+        to each component to store the name of the component.
+    """
+
+    allspice_client.logger.debug(
+        f"Listing components in {dsn_path=} from {repository.get_full_name()} on {ref=}"
+    )
+
+    # Get the generated JSON for the schematic.
+    dsn_json = _fetch_generated_json(repository, dsn_path, ref)
+    pages = dsn_json["pages"]
+    components = []
+
+    for page in pages:
+        for component in page["components"].values():
+            component["_name"] = component["name"]
+            for attribute in component["attributes"].values():
+                component[attribute["name"]] = attribute["value"]
+            components.append(component)
+
+    components = _combine_multi_symbol_components(components)
+
+    return components
+
+
+def _combine_multi_symbol_components(components: list[dict[str, str]]) -> list[dict[str, str]]:
```

Comment:
Since this is a known issue with Altium as well at the moment, can we separate this into a different PR? I think the OrCad changes in this PR are good to go, but I have some feedback/concerns regarding the multi-symbol component logic that will block the OrCad work.

Namely:
- I'm not sure what value creating `components_by_footprint`s adds currently. You can have multiple multi-symbol components that use the same footprint. `PCB_FOOTPRINT_ATTR_NAME` is just a footprint name, not a unique id. I think it could be a helpful assertion to check that symbols we "combining" share the same footprint, but in itself it's misleading to consider it something we should group on. 
- That said, without using the footprint we really just have the Part Reference (Designator) to go off of here. I think that's okay for now, but we should have a more robust solution. The IR should provide an indication that the component is a multi-symbol component. That requires changes to the IR generation code. 

My suggestion for now:
- Lets pull the multi-symbol consolidation out of this PR and create a separate one where we can look at the diff in isolation.
- If we want to keep footprint as a dictionary key the logic could be more simple if we used key: `f"{PCB_FOOTPRINT_ATTR_NAME}-{PART_REFERENCE_ATTR_NAME prefix}` then turned that resulting dict into an array. 
- Investigate whether there is more information we can include in the IR to indicate a symbol should be grouped.

Reply:
I've removed it for now. 

Either way, I'll try to explain the logic behind this. We can deduce a component is actually a symbol of a multi-symbol component if they are part of the same footprint and have the same designator but with a letter appended to it. So this function first groups components by the footprint, then if a footprint has multiple components, checks each component to see if its designator follows that pattern. If it does, we take it as a multi-symbol component and combine all of them into one, using the prefix as the designator instead.

So:

> You can have multiple multi-symbol components that use the same footprint

This would be handled correctly, since we're once again grouping by the prefix. 

> in itself it's misleading to consider it something we should group on.

This code wouldn't, unless the prefix matches too, and the components list would be unchanged in that case. I think the problem here is more that this code is hard to read and understand, and I should've added a few more comments and thought about the naming. It's some consolation that we can verify this works "correctly" by comparing the test case CSV file and the design itself, but that is very manual and difficult to do (there are 840 components 😓)

> If we want to keep footprint as a dictionary key the logic could be more simple if we used key: f"{PCB_FOOTPRINT_ATTR_NAME}-{PART_REFERENCE_ATTR_NAME prefix} then turned that resulting dict into an array.

There is a (very minor) advantage to doing it the way we've done it here: if there is only one component in a footprint, we skip checking the designators. After going through a few examples it looks like that's a pretty rare case, so I don't think it's worth it since it seems to be harder to understand.

> Investigate whether there is more information we can include in the IR to indicate a symbol should be grouped.

This ultimately is the reason why I think removing it for now makes sense. My original reasoning was that it seems simple enough and is a *lot* better for COGS computation, as COGS is simply wrong when a multi-symbol component gets treated as n components. But I'm also unsure if the logic I've used here is airtight, and I would rather we didn't have it and documented the issue than run into a subtle bug later on.